# List Comprehensions Lab

Complete the following set of exercises to solidify your knowledge of list comprehensions.

In [3]:
import os
import numpy as np
import pandas as pd

### 1. Use a list comprehension to create and print a list of consecutive integers starting with 1 and ending with 50.

In [4]:
lst_1 = [x for x in range(1,51)]
print(lst_1)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50]


### 2. Use a list comprehension to create and print a list of even numbers starting with 2 and ending with 200.

In [7]:
lst_2 = [y for y in range(2,201,2)]
print(lst_2)

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182, 184, 186, 188, 190, 192, 194, 196, 198, 200]


### 3. Use a list comprehension to create and print a list containing all elements of the 10 x 4 Numpy array below.

In [55]:
a = np.array([[0.84062117, 0.48006452, 0.7876326 , 0.77109654],
       [0.44409793, 0.09014516, 0.81835917, 0.87645456],
       [0.7066597 , 0.09610873, 0.41247947, 0.57433389],
       [0.29960807, 0.42315023, 0.34452557, 0.4751035 ],
       [0.17003563, 0.46843998, 0.92796258, 0.69814654],
       [0.41290051, 0.19561071, 0.16284783, 0.97016248],
       [0.71725408, 0.87702738, 0.31244595, 0.76615487],
       [0.20754036, 0.57871812, 0.07214068, 0.40356048],
       [0.12149553, 0.53222417, 0.9976855 , 0.12536346],
       [0.80930099, 0.50962849, 0.94555126, 0.33364763]])

[[0.84062117 0.48006452 0.7876326  0.77109654]
 [0.44409793 0.09014516 0.81835917 0.87645456]
 [0.7066597  0.09610873 0.41247947 0.57433389]
 [0.29960807 0.42315023 0.34452557 0.4751035 ]
 [0.17003563 0.46843998 0.92796258 0.69814654]
 [0.41290051 0.19561071 0.16284783 0.97016248]
 [0.71725408 0.87702738 0.31244595 0.76615487]
 [0.20754036 0.57871812 0.07214068 0.40356048]
 [0.12149553 0.53222417 0.9976855  0.12536346]
 [0.80930099 0.50962849 0.94555126 0.33364763]]


In [51]:
lst_3 = np.array([[np.random.rand() for x in range(4)] for y in range(10)])
print(lst_3)

[[0.6785605  0.68168207 0.52696315 0.66115403]
 [0.04919574 0.20691756 0.88900168 0.13093142]
 [0.05682119 0.04472762 0.44825219 0.88988588]
 [0.28407722 0.17378269 0.79392853 0.72875356]
 [0.01280583 0.31116913 0.52540421 0.75187337]
 [0.10046606 0.71219204 0.55356625 0.58158411]
 [0.53249658 0.03841526 0.0031041  0.43290427]
 [0.14859152 0.54094702 0.83022186 0.71943017]
 [0.34765871 0.89808749 0.389307   0.42651892]
 [0.78997977 0.13279754 0.02196108 0.91035835]]


### 4. Add a condition to the list comprehension above so that only values greater than or equal to 0.5 are printed.

In [54]:
lst_4 = np.array([y for x in lst_3 for y in x if y>=.5])
print(lst_4)

[0.6785605  0.68168207 0.52696315 0.66115403 0.88900168 0.88988588
 0.79392853 0.72875356 0.52540421 0.75187337 0.71219204 0.55356625
 0.58158411 0.53249658 0.54094702 0.83022186 0.71943017 0.89808749
 0.78997977 0.91035835]


### 5. Use a list comprehension to create and print a list containing all elements of the 5 x 2 x 3 Numpy array below.

In [14]:
b = np.array([[[0.55867166, 0.06210792, 0.08147297],
        [0.82579068, 0.91512478, 0.06833034]],

       [[0.05440634, 0.65857693, 0.30296619],
        [0.06769833, 0.96031863, 0.51293743]],

       [[0.09143215, 0.71893382, 0.45850679],
        [0.58256464, 0.59005654, 0.56266457]],

       [[0.71600294, 0.87392666, 0.11434044],
        [0.8694668 , 0.65669313, 0.10708681]],

       [[0.07529684, 0.46470767, 0.47984544],
        [0.65368638, 0.14901286, 0.23760688]]])

In [81]:
lst_5 = np.random.random((5,2,3))
print(lst_5)

[[[0.26521723 0.29219423 0.60766795]
  [0.49448474 0.49411711 0.86358014]]

 [[0.81312688 0.83912218 0.28597201]
  [0.18043799 0.65414272 0.52505572]]

 [[0.46693497 0.20427674 0.57326751]
  [0.8604098  0.78033093 0.10536864]]

 [[0.63401215 0.27478901 0.4216823 ]
  [0.35378253 0.67291889 0.29193706]]

 [[0.62590285 0.46979336 0.14246464]
  [0.23135735 0.19632439 0.15971444]]]


### 5. Add a condition to the list comprehension above so that the last value in each subarray is printed, but only if it is less than or equal to 0.5.

In [86]:
lst_5_b = np.array([y for x in lst_3 for y in x if y>=.5])
print(lst_5_b)

[0.6785605  0.68168207 0.52696315 0.66115403 0.88900168 0.88988588
 0.79392853 0.72875356 0.52540421 0.75187337 0.71219204 0.55356625
 0.58158411 0.53249658 0.54094702 0.83022186 0.71943017 0.89808749
 0.78997977 0.91035835]


### 6. Use a list comprehension to select and print the names of all CSV files in the */data* directory.

### 7. Use a list comprehension and the Pandas `read_csv` and `concat` methods to read all CSV files in the */data* directory and combine them into a single data frame. Display the top 10 rows of the resulting data frame.

### 8. Use a list comprehension to select and print the column numbers for columns from the data set whose median is less than 0.48.

### 9. Use a list comprehension to add a new column (20) to the data frame whose values are the values in column 19 minus 0.1. Display the top 10 rows of the resulting data frame.

### 10. Use a list comprehension to extract and print all values from the data set that are between 0.7 and 0.75.